File: pico/hinge_manager.py
from clock import clock

OPEN_DELTA = 2.0  # seconds
CLOSED_DELTA = 2.0  # seconds


class HingeManager:
    def __init__(self):
        self.pin_open = False
        self.pin_open_since = clock.get_time()
        self.pin_closed_since = clock.get_time()
        self.was_open_for_some_time = False

    def is_open_for_some_time(self):
        return self.pin_open and (self.pin_open_since + OPEN_DELTA < clock.get_time())

    def is_closed_for_some_time(self):
        return (not self.pin_open) and (
            self.pin_closed_since + CLOSED_DELTA < clock.get_time()
        )

    def tick(self, door_is, door_should_be, hinge_open):
        reset_timer = False
        was_open = self.pin_open
        self.pin_open = hinge_open

        if was_open != hinge_open:
            if hinge_open:
                self.pin_open_since = clock.get_time()
            else:
                self.pin_closed_since = clock.get_time()

        if self.is_open_for_some_time():
            self.was_open_for_some_time = True
            reset_timer = True
        if self.is_closed_for_some_time() and self.was_open_for_some_time:
            self.was_open_for_some_time = False
            door_is = True
            reset_timer = True

        return door_is, door_should_be, reset_timer

def hinge_example():
    door_is_open = door_should_open = hinge_open = False
    
    door_is_open, door_should_open, reset_timer = hinge_manager.tick(
        door_is_open, door_should_open, hinge_open
    )

File: pico-c/src/hinge_manager.cpp
#include "hinge_manager.hpp"

HingeManager::HingeManager() : pin_open(false), was_open_for_some_time(false)
{
    pin_open_since = get_absolute_time();
    pin_closed_since = get_absolute_time();
}

bool HingeManager::is_open_for_some_time() const
{
    return pin_open && absolute_time_diff_us(pin_open_since, get_absolute_time()) > OPEN_DELTA * 1e6;
}

bool HingeManager::is_closed_for_some_time() const
{
    return !pin_open && absolute_time_diff_us(pin_closed_since, get_absolute_time()) > CLOSED_DELTA * 1e6;
}

std::tuple<bool, bool, bool> HingeManager::tick(bool door_is, bool door_should_be, bool hinge_open)
{
    bool reset_timer = false;
    bool was_open = pin_open;
    pin_open = hinge_open;

    if (was_open != hinge_open)
    {
        if (hinge_open)
        {
            pin_open_since = get_absolute_time();
        }
        else
        {
            pin_closed_since = get_absolute_time();
        }
    }

    if (is_open_for_some_time())
    {
        was_open_for_some_time = true;
        reset_timer = true;
    }

    if (is_closed_for_some_time() && was_open_for_some_time)
    {
        was_open_for_some_time = false;
        door_is = true;
        reset_timer = true;
    }

    return {door_is, door_should_be, reset_timer};
}
